### PR TITLE
ceph-dencoder: RGW - Add missing types

### PR DESCRIPTION
**Pull request by NitzanMordhai**
Currently, ceph-dencoder lacks certain rgw types, preventing us from accurately checking the ceph corpus for encode-decode mismatches. This pull request aims to address this issue by adding the missing types to ceph-dencoder.

To successfully incorporate these types into ceph-dencoder, we need to introduce the necessary `dump` and `generate_test_instances` functions that was missing in some types. These functions are essential for proper encode and decode of the added types.

This PR will enhance the functionality of ceph-dencoder by including the missing types, enabling a comprehensive analysis of encode-decode consistency. With the addition of these types, we can ensure the robustness and correctness of the ceph corpus.

This update will significantly contribute to improving the overall reliability and accuracy of ceph-dencoder. It allows for a more comprehensive assessment of the encode-decode behavior, leading to enhanced data integrity and stability within the ceph ecosystem.

Fixes: https://tracker.ceph.com/issues/61788

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by NitzanMordhai** [comment]
jenkins test make check

**Comment by NitzanMordhai** [comment]
jenkins retest this please

File: src/cls/rgw/cls_rgw_ops.h [hunk 1]
```diff
@@ -527,6 +527,19 @@ struct rgw_cls_obj_remove_op {
     decode(keep_attr_prefixes, bl);
     DECODE_FINISH(bl);
   }
+
+  void dump(ceph::Formatter *f) const {
+    f->open_array_section("keep_attr_prefixes_list");
+    for (auto iter = keep_attr_prefixes.begin(); iter != keep_attr_prefixes.end(); ++iter) {
+      f->dump_string("keep_attr_prefix", *iter);
+    }
+    f->close_section();
```
Comment (cbodley) [inline src/cls/rgw/cls_rgw_ops.h:536]: the helpers in src/common/ceph_json.h support `std::list<T>` already:
```c++
encode_json("keep_attr_prefixes", keep_attr_prefixes, f);
```

File: src/cls/rgw/cls_rgw_ops.h [hunk 2]
```diff
@@ -728,6 +816,26 @@ struct rgw_cls_bi_list_ret {
...
+
+  void dump(ceph::Formatter *f) const {
+    f->dump_bool("is_truncated", is_truncated);
+    f->open_array_section("entries");
+    for (std::list<rgw_cls_bi_entry>::const_iterator iter = entries.begin(); iter != entries.end(); ++iter) {
+      f->open_object_section("entry");
+      iter->dump(f);
+      f->close_section();
+    }
+    f->close_section();
```
Comment (cbodley) [inline src/cls/rgw/cls_rgw_ops.h:828]:
```c++
encode_json("entries", entries, f);
```

File: src/rgw/driver/rados/rgw_zone.h [hunk 3]
```diff
@@ -94,6 +94,7 @@ class RGWSystemMetaObj {
 
   void dump(Formatter *f) const;
   void decode_json(JSONObj *obj);
+  //static void generate_test_instances(std::list<RGWSystemMetaObj*>& o);
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_zone.h:97]: ?

File: src/rgw/rgw_cors.cc [hunk 4]
```diff
@@ -40,6 +40,30 @@ void RGWCORSRule::dump_origins() {
...
+{
+  f->open_object_section("CORSRule");
+  f->dump_string("ID", id);
+  f->dump_unsigned("MaxAgeSeconds", max_age);
+  f->dump_unsigned("AllowedMethod", allowed_methods);
+  f->open_array_section("AllowedOrigin");
+  for (auto& origin : allowed_origins) {
+    f->dump_string("Origin", origin);
+  }
+  f->close_section();
```
Comment (cbodley) [inline src/rgw/rgw_cors.cc:53]: does encode_json work for these lists too?

File: src/tools/ceph-dencoder/rgw_types.h [hunk 5]
```diff
@@ -147,4 +186,56 @@ TYPE(rgw_data_sync_status)
...
+
+
+#include "rgw_user.h"
+TYPE(RGWUID)
+
+#include "rgw_user_types.h"
+TYPE(rgw_user)
+
+//#include "rgw/driver/rados/rgw_zone.h"
+//TYPE_FEATUREFUL_NONDETERMINISTIC(RGWSystemMetaObj)
```
Comment (cbodley) [inline src/tools/ceph-dencoder/rgw_types.h:240]: why commented out?

**Review by cbodley — COMMENTED** [review]
thanks! we were working on some of this in https://tracker.ceph.com/issues/54054 but didn't cover anything

**Comment by cbodley** [comment]
a 'make check' failure
```
**** reencode of /home/jenkins-build/build/workspace/ceph-pull-requests/ceph-object-corpus/archive/15.0.0-539-g191ab33faf/objects/rgw_placement_rule/276cee7e7a31943b9d27987b75241f8d resulted in a different dump ****
3c3
<     "storage_class": "STANDARD"
---
>     "storage_class": ""
```

File: src/cls/rgw/cls_rgw_ops.h (hunk 1 above)
Comment (NitzanMordhai) [reply src/cls/rgw/cls_rgw_ops.h:536]: ok, missed that encode_json.. Changing it

File: src/cls/rgw/cls_rgw_ops.h (hunk 2 above)
Comment (NitzanMordhai) [reply src/cls/rgw/cls_rgw_ops.h:828]: Done

File: src/tools/ceph-dencoder/rgw_types.h (hunk 5 above)
Comment (NitzanMordhai) [reply src/tools/ceph-dencoder/rgw_types.h:240]: Abstract class, need to remove

File: src/rgw/driver/rados/rgw_zone.h (hunk 3 above)
Comment (NitzanMordhai) [reply src/rgw/driver/rados/rgw_zone.h:97]: removed

File: src/rgw/rgw_cors.cc (hunk 4 above)
Comment (NitzanMordhai) [reply src/rgw/rgw_cors.cc:53]: Yes, changed

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
jenkins test make check

**Comment by NitzanMordhai** [comment]
**** reencode of /home/jenkins-build/build/workspace/ceph-pull-requests/ceph-object-corpus/archive/15.0.0-539-g191ab33faf/objects/rgw_placement_rule/276cee7e7a31943b9d27987b75241f8d resulted in a different dump ****
3c3
<     "storage_class": "STANDARD"
---
>     "storage_class": ""

fixed the dump function to use the get storage class

**Comment by ivancich** [comment]
@NitzanMordhai -- this PR is different than others I've worked with in that your branch is not on your own fork of ceph/ceph but within ceph/ceph itself. I don't know if that's proper. @cbodley -- what are your thoughts?

**Comment by NitzanMordhai** [comment]
> @NitzanMordhai -- this PR is different than others I've worked with in that your branch is not on your own fork of ceph/ceph but within ceph/ceph itself. I don't know if that's proper. @cbodley -- what are your thoughts?

Yes, I made a mistake with those PRs, and it won't happen again, I talked to @rzarzynski, and it looks like there is no easy way to change the base repository without closing and reopening new pr, which means losing all the comments.

**Comment by ivancich** [comment]
Passed QE:
* https://pulpito.ceph.com/ivancich-2023-09-06_18:12:39-rgw-wip-eric-testing-1-distro-default-smithi/
* https://pulpito.ceph.com/ivancich-2023-09-06_18:12:09-rgw-wip-eric-testing-1-distro-default-smithi/

**Pull request merged**